File: src/main/java/com/ragin/bdd/cucumber/utils/RESTCommunicationUtils.java
package com.ragin.bdd.cucumber.utils;

import com.ragin.bdd.cucumber.core.ScenarioStateContext;
import io.cucumber.datatable.DataTable;
import java.util.List;
import java.util.Map;
import org.junit.Assert;
import org.springframework.http.HttpHeaders;

/**
 * Utils class for common REST communication methods
 */
public final class RESTCommunicationUtils {
    private RESTCommunicationUtils() {}

    /**
     * Create HTTP header
     *
     * @param addAuthorisation  true = add BearerToken | false = no Authorization header
     * @return  default Headers
     */
    public static HttpHeaders createHTTPHeader(final boolean addAuthorisation) {
        final HttpHeaders headers = new HttpHeaders();
        headers.add("Accept", "application/json");
        headers.add("Accept-Language", "en");
        headers.add("Content-Type", "application/json");

        if (addAuthorisation) {
            headers.add("Authorization", "Bearer " + ScenarioStateContext.current().getBearerToken());
        }

        return headers;
    }

    /**
     * Prepare dynamic URL with data from datatable to exchange the dynamic values
     *
     * @param dataTable     DataTable from Cucumber file
     * @return              path with replaced values
     */
    public static String prepareDynamicURLWithDataTable(final DataTable dataTable) {
        // Prepare request
        String path = ScenarioStateContext.current().getUriPath();

        // assert that given for path and body was previously done
        Assert.assertNotNull("No given path found", path);

        // Read datatable
        final List<Map<String, String>> dataTableRowList = dataTable.asMaps(String.class, String.class);

        for (Map<String, String> stringStringMap : dataTableRowList) {
            // Try to resolve value from context map
            String uriValue = ScenarioStateContext.current().getScenarioContextMap().get(stringStringMap.get("URI Values"));
            // If context map knows nothing about the value, use value directly
            if (uriValue == null) {
                uriValue = stringStringMap.get("URI Values");
            }
            // replace path with URI key and URI value
            path = path.replace(
                    "{" + stringStringMap.get("URI Elements") + "}",
                    uriValue
            );
        }

        return path;
    }
}

File: src/main/kotlin/com/ragin/bdd/cucumber/utils/RESTCommunicationUtils.kt
package com.ragin.bdd.cucumber.utils

import com.ragin.bdd.cucumber.BddLibConstants.BDD_URI_ELEMENTS
import com.ragin.bdd.cucumber.BddLibConstants.BDD_URI_VALUES
import com.ragin.bdd.cucumber.core.ScenarioStateContext.bearerToken
import com.ragin.bdd.cucumber.core.ScenarioStateContext.headerValues
import com.ragin.bdd.cucumber.core.ScenarioStateContext.scenarioContextMap
import com.ragin.bdd.cucumber.core.ScenarioStateContext.uriPath
import io.cucumber.datatable.DataTable
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpHeaders.ACCEPT
import org.springframework.http.HttpHeaders.CONTENT_TYPE
import org.springframework.http.MediaType.APPLICATION_JSON_VALUE
import kotlin.test.assertNotNull

/**
 * Utils class for common REST communication methods
 */
object RESTCommunicationUtils {
    /**
     * Create HTTP header
     *
     * @param addAuthorisation  true = add BearerToken | false = no Authorization header
     * @return  default Headers
     */
    @JvmStatic
    fun createHTTPHeader(addAuthorisation: Boolean): HttpHeaders {
        val headers = HttpHeaders()
        if (addAuthorisation && headerValues[HttpHeaders.AUTHORIZATION] == null) {
            headers.add(HttpHeaders.AUTHORIZATION, "Bearer $bearerToken")
        }

        // set user specific headers if present
        if (headerValues.isNotEmpty()) {
            for (headerName in headerValues.keys) {
                headers.add(headerName, headerValues[headerName])
            }
        }

        if (! headerValues.containsKey(CONTENT_TYPE)) {
            headers.add(CONTENT_TYPE, APPLICATION_JSON_VALUE)
        }

        if (! headerValues.containsKey(ACCEPT)) {
            headers.add(ACCEPT, APPLICATION_JSON_VALUE)
        }

        return headers
    }

    /**
     * Prepare dynamic URL with data from datatable to exchange the dynamic values
     *
     * @param dataTable     DataTable from Cucumber file
     * @return              path with replaced values
     */
    @JvmStatic
    fun prepareDynamicURLWithDataTable(dataTable: DataTable): String {
        // Prepare request
        var path = uriPath

        // assert that given for path and body was previously done
        assertNotNull(
            actual = path,
            message = "No given path found"
        )

        // Read datatable
        val dataTableRowList = dataTable.asMaps(String::class.java, String::class.java)
        for (stringStringMap in dataTableRowList) {
            // Try to resolve value from context map
            if (! stringStringMap[BDD_URI_VALUES].isNullOrEmpty()) {
                var uriValue = scenarioContextMap[stringStringMap[BDD_URI_VALUES]]
                // If context map knows nothing about the value, use value directly
                if (uriValue == null) {
                    uriValue = stringStringMap[BDD_URI_VALUES]
                }
                // replace path with URI key and URI value
                path = path.replace(
                    oldValue = "{${stringStringMap[BDD_URI_ELEMENTS]}}",
                    newValue = uriValue!!
                )
            }
        }
        return path
    }
}
